### PR TITLE
Add file type detection

### LINKED ISSUE #240: Feature Request: File Type Detection
I have a script generating random names for images based on the MD5 hash of their content. As a result these files have no extension. When I run imagemin on this folder it says "Minified 0 images"

Could something like [file-type](https://www.npmjs.com/package/file-type) be used to allow for minification of improperly named files?

### PR DESCRIPTION
**Pull request by stevendesu**
Fixes #240

### CODE REVIEW

**Comment by sindresorhus** [comment]
It should still check file extensions, but check the content file type if no file extension matches.

File: index.js [hunk 1]
```diff
@@ -45,7 +46,7 @@ module.exports = (plugins, opts) => {
 		verbose: process.argv.indexOf('--verbose') !== -1
 	}, opts);
 
-	const validExts = ['.jpg', '.jpeg', '.png', '.gif', '.svg'];
+	const validMimes = ['image/jpeg', 'image/pjpeg', 'image/png', 'image/gif', 'image/svg+xml'];
```
Comment (sindresorhus) [inline index.js]: `file-type` would never return `image/pjpeg`.

**Comment by sindresorhus** [comment]
And needs a test.

**Comment by stevendesu** [comment]
**It should still check file extensions:**

I'll stick extension checks back in. I figured if we were using magic numbers to check it eliminated the need, and would also allow for properly minifying images with the _**incorrect**_ extension (if someone performed a mass rename or similar operation and in doing so changed a PNG to a `.jpg`)

Maybe the options could be extended to include which type checks to perform (since `file-type` can cause noticeable although not unmanageable overhead for large numbers of files). Something like:

    gulp.src(...)
         .pipe(imagemin({
            validityChecks: ['extension', 'mimeType']
        })

**`file-type` would never return `image/pjpeg`:**

`file-type` might add `image/pjpeg` in the future, so it's good to recognize that most (if not all) imagemin plugins support it.

I think a bigger concern, as I was just looking at the output values of `file-type`, is that it fails to detect SVG. I'll have to extend my solution with `is-svg` or similar

**And needs a test:**

I performed smoke testing to ensure my syntax and usage of `file-type` was correct, but I'll do a full end-to-end test this weekend running gulp on a directory with multiple files (some properly named, some improperly named, some without extensions, some not even images) and see what happens.

**Travis-CI test failed:**

Looks like Travis threw two error, that both look like linting issues:

 1. 2:7   path is assigned a value but never used.  no-unused-vars
 2. 66:35  fs is not defined.                        no-undef

Both easily resolved and I'll update them (along with adding `is-svg` and putting extension checks back in) tonight.

**Comment by stevendesu** [comment]
Actually realized it wouldn't take too long to update, so went ahead and updated. Also set up a local project where I included this version of `gulp-imagemin` and tested it on 16 files:

    actual-gif.gif
    actual-jpeg.jpg
    actual-jpg.jpg
    actual-png.png
    actual-svg.svg
    fake-gif.gif
    fake-jpeg.jpeg
    fake-jpg.jpg
    fake-png.png
    fake-svg.svg
    gif-noext
    jpeg-noext
    jpg-noext
    png-noext
    svg-noext
    nonimage-noext

It properly minified what it could and passed through what it couldn't minify untouched. There were no errors and I was pleased with the results

Failed a number of lint checks, though, so going to make one more commit here shortly

**Comment by sindresorhus** [comment]
> and would also allow for properly minifying images with the incorrect extension (if someone performed a mass rename or similar operation and in doing so changed a PNG to a .jpg)

That would still be supported. Reread my comment. I'm requesting checking extension first, then if it doesn't match, check the content.

> Maybe the options could be extended to include which type checks to perform (since file-type can cause noticeable although not unmanageable overhead for large numbers of files)

Nah

> Therefore renaming a PNG file to .jpg it still detected it as PNG and properly minified it.

That's not a use-case I care about.

> Since imagemin already checks file types before attempting to minify I don't actually see the reason for gulp-imagemin to bother checking for valid types in the first place

I still prefer to have such check here.

> file-type might add image/pjpeg in the future

I don't add stuff because of maybes.

**Comment by stevendesu** [comment]
>I still prefer to have such check here.

I don't understand why you prefer to have additional overhead for no gain in functionality, but whatever floats your boat 😄 

>I don't add stuff because of maybes.

The reason I feel this should stay is that there's two options:

 1. `file-type` never adds an `image/pjpeg` mime type:

In this case, there's no negative impact. Your code continues to function and no one knows the difference. It's benign code sitting there hurting nothing.

 2. `file-type` **does** add an `image/jpeg` mime type:

In this case, your plugin will suddenly stop working for progressively-encoded JPEG files. People will get confused and several issues will be opened about how your plugin is suddenly busted. I exaggerate, of course, because the issue will only be present on progressively-encoded JPEG files _**with no file extension**_, but the point remains: checking for `image/pjpeg` breaks nothing and **not** checking for it could **potentially** break something in the future.

**Comment by stevendesu** [comment]
Mind me asking why this hasn't been accepted 2 weeks later? If it's just the `image/pjpeg` thing then if I remove that will it get merged?

**Comment by sindresorhus** [comment]
> file-type does add an image/pjpeg mime type:

I maintain `file-type` and can say with certainty that it will never be added. That mime-type might make sense when sent from a server, but not when detecting the file type.

> Mind me asking why this hasn't been accepted 2 weeks later? 

Simply because I also have a life outside GitHub: https://twitter.com/sindresorhus/status/817076702891429889 :)

> If it's just the image/pjpeg thing then if I remove that will it get merged?

Yes

File: index.js [hunk 2]
```diff
@@ -62,7 +66,23 @@ module.exports = (plugins, opts) => {
 			return;
 		}
 
-		if (validExts.indexOf(path.extname(file.path).toLowerCase()) === -1) {
+		if (path.extname(file.path) === '') {
+			// File has no extension -- fall back to file type detection
```
Comment (sindresorhus) [inline index.js]: This comment should be above the if-statement.

File: index.js [hunk 3]
```diff
@@ -62,7 +66,23 @@ module.exports = (plugins, opts) => {
...
+			const fileBuffer = fs.readFileSync(file.path);
```
Comment (sindresorhus) [inline index.js]: You already have the data in `file.contents`. No point in re-reading it.

File: index.js [hunk 4]
```diff
@@ -62,7 +66,23 @@ module.exports = (plugins, opts) => {
...
+			const type = fileType(fileBuffer);
+			if (type && validMimes.indexOf(type.mime) === -1) {
```
Comment (sindresorhus) [inline index.js]: Use `String#includes`

File: index.js [hunk 5]
```diff
@@ -62,7 +66,23 @@ module.exports = (plugins, opts) => {
...
+				if (!isSvg(fileBuffer)) {
+					if (opts.verbose) {
+						gutil.log(`gulp-imagemin: Skipping unsupported image ${chalk.blue(file.relative)}`);
+					}
+
+					cb(null, file);
+					return;
+				}
+			}
+		} else if (validExts.indexOf(path.extname(file.path).toLowerCase()) === -1) {
```
Comment (sindresorhus) [inline index.js]: Use `String#includes`

File: index.js [hunk 6]
```diff
@@ -62,7 +66,23 @@ module.exports = (plugins, opts) => {
 			return;
 		}
 
-		if (validExts.indexOf(path.extname(file.path).toLowerCase()) === -1) {
+		if (path.extname(file.path) === '') {
+			// File has no extension -- fall back to file type detection
+			const fileBuffer = fs.readFileSync(file.path);
+			const type = fileType(fileBuffer);
+			if (type && validMimes.indexOf(type.mime) === -1) {
+				// file-type does not (currently) detect SVG files
```
Comment (sindresorhus) [inline index.js]: Drop `(currently)`

File: index.js [hunk 7]
```diff
@@ -62,7 +66,23 @@ module.exports = (plugins, opts) => {
 			return;
 		}
 
-		if (validExts.indexOf(path.extname(file.path).toLowerCase()) === -1) {
+		if (path.extname(file.path) === '') {
```
Comment (sindresorhus) [inline index.js]: Use `file.extname`

File: index.js (hunk 5 above)
Comment (sindresorhus) [inline index.js]: Use `file.extname`

**Comment by stevendesu** [comment]
What version of NodeJS would you like to be compatible with? I have `v4.7.1` installed locally and most people have `v5` or `v6` installed. In all of these versions the following error is thrown:

```
/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/index.js:84
                } else if (validExts.includes(path.extname(file.path).toLowerCase())) {
                                     ^

TypeError: validExts.includes is not a function
    at DestroyableTransform.through.obj.imagemin.buffer.then.data.catch.percent (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/index.js:84:24)
    at DestroyableTransform._transform (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/node_modules/through2-concurrent/through2-concurrent.js:30:15)
    at DestroyableTransform.Transform._read (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/node_modules/through2-concurrent/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:159:10)
    at DestroyableTransform.Transform._write (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/node_modules/through2-concurrent/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:147:83)
    at doWrite (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/node_modules/through2-concurrent/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:347:64)
    at writeOrBuffer (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/node_modules/through2-concurrent/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:336:5)
    at DestroyableTransform.Writable.write (/home/sbarnett/gulp-imagemin-test/node_modules/gulp-imagemin/node_modules/through2-concurrent/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:274:11)
    at write (/home/sbarnett/gulp-imagemin-test/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:623:24)
    at flow (/home/sbarnett/gulp-imagemin-test/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:632:7)
    at DestroyableTransform.pipeOnReadable (/home/sbarnett/gulp-imagemin-test/node_modules/gulp/node_modules/vinyl-fs/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:664:5)
```

You can work around this with the `--harmony_array_includes` flag, however most people probably wouldn't accept this as they'd rather just type `gulp` than have to type `gulp --harmony_array_includes`

**Comment by sindresorhus** [comment]
Ugh, you're right. I always forget the Array version of includes is not yet available. Never mind that change then.

**Comment by sindresorhus** [comment]
In case you missed it. This is waiting for a test and readme update before being merged, as commented in https://github.com/sindresorhus/gulp-imagemin/pull/241#issuecomment-268855457 and https://github.com/sindresorhus/gulp-imagemin/pull/241#pullrequestreview-15384735.

**Comment by stevendesu** [comment]
I've got other priorities on my plate right now and I've done the main part. I'll leave the PR open in case anyone else wants to add the tests and update the docs. In the mean time, I've found a work-around for the specific issue I was having (#240) -- I just have a minor pre-processing and cleanup script which runs before/after gulp to append extensions and then remove them

**Pull request closed without merging**